doc/source/development/window.rst

```diff
@@ -0,0 +1,258 @@
+           return self.agg(self.values[start:stop])
+
+
+A *kernel* is an interface that implements methods to update its internal state
+as well as a :meth:`finalize` method to return the current value of the
+aggregation. Its constructor must take no arguments.
+
+An example implementation of a :class:`Sum` kernel would look similar to the
+following:
+
+.. code-block:: python
+
+   class AggKernel(Generic[Output]):
+       def __init__(self):
+           self.count = 0
+
+       @abc.abstractmethod
+       def finalize(self) -> Optional[Output]:
+           """Return the final value of the aggregation."""
+
+
+   class UnaryAggKernel(Generic[Input1, Output], AggKernel[Output]):
+       @abc.abstractmethod
+       def step(self, value: Optional[Input1]) -> None:
+           """Update the state of the aggregation with `value`."""
+
+
+   class Sum(UnaryAggKernel[Input1, Output]):
+       def __init__(self):
+           super().__init__()
+           self.total: Output = 0
+
+       def step(self, value: Optional[Input1]) -> None:
+           if value is not None:
+               self.count += 1
+               self.total += value
+
+       def invert(self, value: Optional[Input1]) -> None:
+           """Used only in subtractable kernels."""
+           if value is not None:
+               self.count -= 1
+               self.total -= value
+
+       def finalize(self) -> Optional[Output]:
+           if not self.count:
+               return None
+           return self.total
+
+       def combine(self, other: Sum[Input1, Output]) -> None:
+           """Used only in segment tree aggregator."""
+           self.total += other.total
+           self.count += other.count
+
+       @classmethod
+       def make_aggregator(
+           cls, values: ndarray[Input]
+       ) -> AggregationAlgorithm:
+           SubtractableAggregator.check_agg(cls)
+           aggregator = SubtractableAggregator(values, cls())
+           return aggregator
+
+   class BinaryAggKernel(Generic[Input1, Input2, Output], AggKernel[Output]):
+       @abc.abstractmethod
+       def step(
+           self, value1: Optional[Input1], value2: Optional[Input2]
+       ) -> None:
+           ...
+
+   class Covariance(BinaryAggKernel[Input1, Input2, float])
+       def step(
+           self, value1: Optional[Input1], value2: Optional[Input2]
+       ) -> None:
+           if value1 is not None and value2 is not None:
+              ...
+
+       def finalize(self) -> Optional[float]:
+           ...
+
+
+Customization of Window Spans
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+We propose a generic mechanism that allows power users and library authors to
+customize the computation of window boundaries.
+
+.. code-block:: python
+
+   import numpy as np
+   from typing import Sequence, Tuple
+
+   BeginEnd = Tuple[np.ndarray[np.int64], np.ndarray[np.int64]]
+   Displacement = TypeVar("Displacement")
+
+   class Indexer(abc.ABC):
```

Comment:
Is the idea that a power user constructs their `Indexer` class that would be fed into the current `rolling` API?

```
windower = MyIndexer(...)
df.rolling(windower).mean()
```

Reply:
Yes.